controllers/auth.js

```diff
@@ -9,6 +9,91 @@ const {
   USER_DOES_NOT_EXIST_ERROR,
 } = require("../constants/errorMessages");
 
+const googleAuthLogin = (req, res, next) => {
+  const { redirectURL, dev } = req.query;
+  if (dev === "true") {
+    return passport.authenticate("google", {
+      scope: ["email"],
+      state: redirectURL,
+    })(req, res, next);
+  } else {
+    return res.boom.unauthorized("User cannot be authenticated");
+  }
+};
+
+async function handleGoogleLogin(req, res, user, authRedirectionUrl) {
+  const rdsUiUrl = new URL(config.get("services.rdsUi.baseUrl"));
+  try {
+    if (!user.emails || user.emails.length === 0) {
+      throw new Error("Email not found in user data");
+    }
+    const userData = {
+      email: user.emails[0].value,
+      created_at: Date.now(),
+      updated_at: null,
+    };
+
+    const userDataFromDB = await users.fetchUser({ email: userData.email });
+
+    if (userDataFromDB.userExists) {
+      if (userDataFromDB.user.roles?.developer) {
+        const errorMessage = encodeURIComponent("Google login is restricted for developer role.");
+        return res.redirect(`${authRedirectionUrl}?error=${errorMessage}`);
+      }
+    }
+
+    const { userId, incompleteUserDetails } = await users.addOrUpdate(userData);
+
+    const token = authService.generateAuthToken({ userId });
+
+    const cookieOptions = {
+      domain: rdsUiUrl.hostname,
+      expires: new Date(Date.now() + config.get("userToken.ttl") * 1000),
+      httpOnly: true,
+      secure: true,
+      sameSite: "lax",
+    };
+
+    res.cookie(config.get("userToken.cookieName"), token, cookieOptions);
+
+    if (incompleteUserDetails) {
+      authRedirectionUrl = "https://my.realdevsquad.com/new-signup";
+    }
+
+    return res.redirect(authRedirectionUrl);
+  } catch (err) {
+    logger.error(err);
+    return res.boom.unauthorized("User cannot be authenticated");
+  }
```

Comment:
Users are being sent generic error messages. To improve debugging, distinct error messages should be sent for different error scenarios. Some errors might not be authentication failures but server errors, yet they're all returning `401`.

Reply:
I have used 401 Unauthorized in the error handling as most of the errors are related to authentication or issues with the user's Google data (such as missing or unverified emails). Since these are primarily authentication-related failures, I believe a 401 is the most appropriate response. Additionally, the error details are logged for debugging purposes.
Let me know if you would like me to adjust the error handling further or add more specific error messages.

Reply:
While logging helps with debugging, we should still provide more specific error messages to users. Even within authentication failures, there can be different scenarios (invalid email, role restrictions, server errors, etc) that should be communicated differently. This helps both users and front-end developers handle errors appropriately.

Reply:
I have added more specific error messages for users.